crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,564 @@
+    }
+
+    // Do one last dry-run
+    let SimulateTransactionResult {
+        input_objects,
+        output_objects,
+        events,
+        effects,
+        ..
+    } = executor
+        .simulate_transaction(resolved_transaction.clone())
+        .map_err(anyhow::Error::from)?;
+
+    let events = events.map(Into::into);
+    let effects = effects.into();
+
+    let input_objects = input_objects
+        .into_values()
+        .map(Into::into)
+        .collect::<Vec<_>>();
+    let output_objects = output_objects
+        .into_values()
+        .map(Into::into)
+        .collect::<Vec<_>>();
+    let balance_changes = derive_balance_changes(&effects, &input_objects, &output_objects);
+
+    ResolveTransactionResponse {
+        transaction: resolved_transaction.into(),
+        events,
+        effects,
+        balance_changes: Some(balance_changes),
+        input_objects: Some(input_objects),
+        output_objects: Some(output_objects),
+    }
+    .pipe(|response| match accept {
+        AcceptFormat::Json => ResponseContent::Json(response),
+        AcceptFormat::Bcs => ResponseContent::Bcs(response),
+    })
+    .pipe(Ok)
+}
+
+struct NormalizedPackage {
+    #[allow(unused)]
+    package: MovePackage,
+    normalized_modules: BTreeMap<String, normalized::Module>,
+}
+
+fn called_packages(
+    reader: &StateReader,
+    protocol_config: &ProtocolConfig,
+    unresolved_transaction: &UnresolvedTransaction,
+) -> Result<HashMap<ObjectId, NormalizedPackage>> {
+    let binary_config = sui_types::execution_config_utils::to_binary_config(protocol_config);
+    let mut packages = HashMap::new();
+
+    for move_call in unresolved_transaction
+        .ptb
+        .commands
+        .iter()
+        .filter_map(|command| {
+            if let Command::MoveCall(move_call) = command {
+                Some(move_call)
+            } else {
+                None
+            }
+        })
+    {
+        let package = reader
+            .inner()
+            .get_object(&(move_call.package.into()))?
+            .ok_or_else(|| ObjectNotFoundError::new(move_call.package))?
+            .data
+            .try_as_package()
+            .ok_or_else(|| {
+                RestError::new(
+                    axum::http::StatusCode::BAD_REQUEST,
+                    format!("object {} is not a package", move_call.package),
+                )
+            })?
+            .to_owned();
+
+        let normalized_modules = package.normalize(&binary_config).map_err(|e| {
+            RestError::new(
+                axum::http::StatusCode::INTERNAL_SERVER_ERROR,
+                format!("unable to normalize package {}: {e}", move_call.package),
+            )
+        })?;
+        let package = NormalizedPackage {
+            package,
+            normalized_modules,
+        };
+
+        packages.insert(move_call.package, package);
+    }
+
+    Ok(packages)
+}
+
+fn resolve_object_references(
```

Comment:
nit: This is a strange name for this function which now seems to do more than just resolve object references -- it seems to resolve the whole transaction?